src/test/java/com/rocketden/main/service/ProblemServiceTests.java

```diff
@@ -333,4 +361,210 @@ public void getRandomProblemNotFound() {
+
+        ProblemInput problemInput = new ProblemInput(INPUT_NAME, IO_TYPE);
+        problem.addProblemInput(problemInput);
+        problem.setOutputType(IO_TYPE);
+
+        Mockito.doReturn(problem).when(repository).findProblemByProblemId(problem.getProblemId());
+
+        ProblemTestCaseDto testCaseDto = new ProblemTestCaseDto();
+        testCaseDto.setInput(INPUT);
+        testCaseDto.setOutput(OUTPUT);
+
+        ProblemDto updatedProblem = ProblemMapper.toDto(problem);
+        updatedProblem.setTestCases(Collections.singletonList(testCaseDto));
+        updatedProblem.setOutputType(null);
+
+        ApiException exception = assertThrows(ApiException.class, () ->
+                problemService.editProblem(problem.getProblemId(), updatedProblem));
+
+        assertEquals(ProblemError.EMPTY_FIELD, exception.getError());
+    }
+
+    @Test
+    public void editProblemNewProblemInputInvalidatesTestCases() {
+        Problem problem = new Problem();
+        problem.setName(NAME);
+        problem.setDescription(DESCRIPTION);
+        problem.setDifficulty(ProblemDifficulty.MEDIUM);
+
+        ProblemInput problemInput = new ProblemInput(INPUT_NAME, IO_TYPE);
+        problem.addProblemInput(problemInput);
+        problem.setOutputType(IO_TYPE);
+
+        ProblemTestCase testCase = new ProblemTestCase();
+        testCase.setInput(INPUT);
+        testCase.setOutput(OUTPUT);
+        problem.addTestCase(testCase);
+
+        Mockito.doReturn(problem).when(repository).findProblemByProblemId(problem.getProblemId());
+
+        ProblemDto updatedProblem = ProblemMapper.toDto(problem);
+        updatedProblem.getProblemInputs().get(0).setType(IO_TYPE_2);
+
+        ApiException exception = assertThrows(ApiException.class, () ->
+                problemService.editProblem(problem.getProblemId(), updatedProblem));
+
+        assertEquals(ProblemError.INVALID_INPUT, exception.getError());
+    }
+
+    @Test
+    public void deleteProblemSuccess() {
+        Problem problem = new Problem();
+        problem.setName(NAME);
+        problem.setDescription(DESCRIPTION);
+        problem.setDifficulty(ProblemDifficulty.MEDIUM);
+
+        Mockito.doReturn(problem).when(repository).findProblemByProblemId(problem.getProblemId());
+
+        ProblemDto response = problemService.deleteProblem(problem.getProblemId());
+
+        verify(repository).delete(problem);
+
+        assertEquals(problem.getName(), response.getName());
+        assertEquals(problem.getDescription(), response.getDescription());
+        assertEquals(problem.getDifficulty(), response.getDifficulty());
+    }
+
+    @Test
+    public void deleteProblemFailure() {
+        ApiException exception = assertThrows(ApiException.class, () -> problemService.deleteProblem("ZZZ"));
+
+        verify(repository).findProblemByProblemId("ZZZ");
+        assertEquals(ProblemError.NOT_FOUND, exception.getError());
+    }
+
+    @Test
+    public void validateGsonParseable() {
```

Comment:
I'm not sure if I missed it, but do you have a test case where multiple inputs and input types are _correctly_ provided for a problem? You addressed the errors on this, I want to check that `String.split("\n");` is doing its job basically.

Reply:
I think the first three lines here should verify that they work - although were you referring to like a full unit/API test that covers multi-line inputs? If so, I can def add that too, just lmk

Reply:
Ah I see -- I didn't connect in my mind the idea that the method shown only throws an error if it _doesn't_ work, and otherwise just does nothing. Is there any case where that could fail but the exception is not checked, so nothing happens? I guess I'm wondering if you should just stick a `return true` at the end of something and check that is returned.

I don't think this is necessary, so I'm guessing you can just say so and resolve this comment, but I'm leaving this comment anyway just in case.

Reply:
Hm yeah that's a good point, I hadn't thought about a case in which it would fail but not throw an exception. My suspicion is that this scenario wouldn't be possible (since if something bad happens for it to not return true then an exception has to be thrown) so I'll prob leave it as is, but I'm ambivalent either way if you want to unresolved this